### ISSUE TITLE
Extract kerning and shaping tables from fonts?

### ISSUE THREAD

**Issue by mikemorris** [mapbox/node-fontnik#85]
**tl;dr** Can we use ICU or HarfBuzz to extract shaping and kerning tables from a font?

From discussion with @nickidlugash on adding kerning support:

> FreeType only supports kerning via the (rather simple) ‘kern’ table. For the interpretation of kerning data in the (highly sophisticated) ‘GPOS’ table you need a higher-level library like ICU or HarfBuzz since it can be context dependent (this is, the kerning may vary depending on the position within a text string, for example).
http://freetype.org/freetype2/docs/glyphs/glyphs-4.html

- FreeType has [limited support](http://www.freetype.org/freetype2/docs/reference/ft2-base_interface.html#FT_FACE_FLAG_KERNING) for reading kerning tables and running [`FT_Get_Kerning`](http://www.freetype.org/freetype2/docs/reference/ft2-base_interface.html#FT_Get_Kerning) for all possible glyph pairs seems like a terrible idea.
- We'll need to integrate HarfBuzz (and a JavaScript equivalent?) into our rendering engines at some point for [complex text shaping](https://github.com/mapbox/mapbox-gl-js/issues/471), could we handle kerning at this point too?
- Is it possible to use ICU or HarfBuzz to extract shaping and kerning tables from a font at upload time to be stashed on S3 and used later for rendering, rather than having to load the actual font file into HarfBuzz/ICU?

**Comment by springmeyer** [comment]
Good questions, I'll add this tidbit:

```
FreeType can now use the HarfBuzz library to greatly improve the
      auto-hinting of  fonts that  use OpenType features:  Many glyphs
      that are part  of such features but don't have  cmap entries are
      now handled  properly, for  example small caps  or superscripts.
      Define the configuration  macro FT_CONFIG_OPTION_USE_HARFBUZZ to
      activate HarfBuzz support
```

http://sourceforge.net/projects/freetype/files/freetype2/2.5.3/

**Comment by mattdesl** [comment]
Opentype.js has also been working on basic GPOS parsing:
https://github.com/nodebox/opentype.js/issues/19

**Comment by mikemorris** [comment]
Current thinking is to implement something Harfbuzz's [`reference_table`](https://github.com/behdad/harfbuzz/blob/master/src/hb-ft.cc#L311-L336) method to extract and dump the binary SFNT table data while node-fontnik has the font open in FreeType, and write a custom backend for HarfBuzz to read this data by itself rather than needing to open up a font file at shaping time. @lbud is digging into how this will work.

I started working on porting [pomax/A-binary-parser-generator](https://github.com/pomax/A-binary-parser-generator) to [Node.js with Browserify](https://github.com/mapbox/A-binary-parser-generator/tree/node-browserify) today to try extracting and parsing the shaping tables in JavaScript, but I think that approach will end up being slower, more difficult and result in less useful data than just doing this through FreeType.

**Comment by felipesanches** [comment]
Ok... I've been reading the specs and getting used to the source code and now things are a bit clearer.

For starters, the proposed setKerningValue method will probably require the user to pass a 2-dimensional vector (dx, dy) because the kerning adjustment can also happen in the Y-axis.

We may accept both types of data, defaulting dy to zero if a single value is passed instead of the pair of offsets.

**Comment by felipesanches** [comment]
In order to implement the makeGposTable function (that encodes the GPOS table) we need access to the relevant data. Given that our current GPOS parser is incomplete, the output GPOS table will be incomplete as well.

**Comment by felipesanches** [comment]
While we may be happy about parsing only a portion of the table, the opposite may not be true. I guess that OTF files with incomplete GPOS tables might annoy the parsers of other programs. Anyone here knows if this is the case?

We may consider improving the parser before actually implementing the encoder.

**Comment by felipesanches** [comment]
Even if we decide to implement the encoder without improving the parser first, the current parsing code will have to be modified. That's because currently it exposes the kerning values by dinamically creating a getKerningValue function that operates on the parsed data that is only accessible to this function because it is trapped in its bound subtables variable (by creating a closure).

In order to output the table we need direct access to the parsed data.

**Comment by felipesanches** [comment]
I just posted a few comments here while I actually intended to post them to issue #95. I have just deleted them from here and posted them to the correct issue. Sorry.
